### DISCUSSION TITLE
[RFC] Toolpad navigation surface revamp (WIP)

### DISCUSSION THREAD

**Discussion by Janpot** [mui/toolpad#2369] [Early feedback]
### What's the problem? 🤔

The left sidebar is designated for all tasks related to navigating the constituents of the project. As we're identifying more parts to be included there, it's getting cluttered. This RFC proposes a reorganization and new UI paradigm to solve the clutter. 

### What are the requirements? ❓

* Be less cluttered
* Be intuitive
* Focuses on a single navigational task at once whenever possible

It needs to be a home for the following:

* Navigation between the different pages of the Toolpad app. Currently part of the left sidebar.
* Component catalog: The place where you start to drag from into the canvas. Currently slides out from under the left sidebar to the right.
* Page hierarchy: The place where get an overview of the elements that make up the canvas. Currently part of the left sidebar
* Page state: The place where you get an overview of all non-ui page state and where you can open editors to create an update it. Currently this is part of the "page options" panel in the right sidebar when no component is selected.
* Global scope explorer: the place where you can inspect all variables and state that is available to the bindings. Currenlty this is displayed on the right side of the binding editor.
* ...?

### What are our options? 💡

_warning: very low-fi wireframes ahead_

The top of the left navigation starts with a header that contains the page title. 

![image](https://github.com/mui/mui-toolpad/assets/2109932/5a80c713-55a2-409d-8eb7-c8a7edeba84d)

Most of the time the focus of the editor lies on the page content. The UI that gives an overview of the different pages adds clutter and takes up screen real estate. That's why this RFC proposes to make this section collapsed. It can be opened by clicking and a list with pages appears to switch between pages. UX-wise this works like a Select, meaning it is always closed unless it's being interacted with.

![image](https://github.com/mui/mui-toolpad/assets/2109932/f9dccf60-ea4f-4014-b938-ac786325d167)

We can add a vertical tabs strip for each of the navigational surfaces:

1. Component catalog:
    The same content as what's currently in the slideout
    
    ![image](https://github.com/mui/mui-toolpad/assets/2109932/7b2457d5-2434-4763-bc69-6c23f63099b3)

    Disadvantages:
     * technically not a "navigational" surface
     * By not being able to show it at the same time as showing the page hierarchy, we take away the option of dragging a new component directly in the hierarchy explorer instead of in the canvas.

    Other options to consider:
      * keep it in a slide out
      * combine it with the hierarchy explorer in a single surface (e.g. add a button on top that opens a popover with the catalog, or split the surface in two and show both surfaces on top of each other)
      * Add a button somewhere else in the UI (e.g. toolbar at the top, like figma) that opens a popover with the component catalog.

2. Page hierarchy
   The same content as what's currently in the page hierarchy explorer
    
    ![Screenshot 2023-07-25 at 13 56 54](https://github.com/mui/mui-toolpad/assets/2109932/8e7e971a-42a5-4514-826a-a008caad2843)

3. Page state:
    The same content as what's now in the right side panel _An update to the UX of this panel will be part of a separate RFC. The point of this RFC is to find the right surface for this panel._
    
    ![Screenshot 2023-07-25 at 13 11 18](https://github.com/mui/mui-toolpad/assets/2109932/5b8996ac-f388-4bdc-a38e-bb16cd66bbf6)

4. Global scope explorer
    The same content as what's now in the binding editor _An update to the UX of this panel will be part of a separate RFC. The point of this RFC is to find the right surface for this panel.

    ![Screenshot 2023-07-25 at 13 12 45](https://github.com/mui/mui-toolpad/assets/2109932/3fe51892-0d09-4a0f-b4b8-650348fa9372)

Discussions:

- Any more navigational panels to consider?
- Where does to component catalog go?
- Detailed UX of page selector (rename, delete create, settings)
- Detailed UI
- Icons?
- ...?

**Comment by apedroferreira** [comment]
I like the idea of selecting an active page from an UI that kind of works like a dropdown, and having the information below be more detailed information about that page. 

At first thought, I think that maybe 2, 3 and 4 should be alternative options of the same view, while the component catalog should be easy to access at any time and separate from those views somehow... Ideally we can also make the component catalog more discoverable, as we have discussed in the past.
Not sure if we can split the left sidebar into component catalog at the top, and page editor below it? It's just another possibility, it all really depends on finding a good design and the available space...

**Comment by bharatkashyap** [comment]
# App-level sidebar
- Pages
- Code components
- Theme 
- Settings 
- ... (Connections?)

## Page section

### Secondary sidebar of collapsible and resizable panels

- Pages
- Page State 
- Actions 
- Page Hierarchy

|   |   |
| ----- | ----- |
| ![Screenshot 2023-08-01 at 1 26 43 AM](https://github.com/mui/mui-toolpad/assets/19550456/20ab5597-b040-45da-a51b-61b941a1ca48) | ![Screenshot 2023-08-01 at 1 26 48 AM](https://github.com/mui/mui-toolpad/assets/19550456/85d55169-1c05-4485-92b1-24f666a10bc9) |

**New feature:** _Page State and Page Hierarchy items allow hovering over them to view their currently evaluated values._

| Tab |  Function |
| ---- | --------- |
| *Pages* |Same as current |
| *Page State*| **Automatic queries** and **Page parameters** are merged to form "Page State" which contains all data that is brought onto the page |
| *Actions* | We rename **Manual queries** to Actions and give them a separate section (everything else remains the same) |
| *Page Hierarchy* | Same as current |

### Create/Edit Page State and Actions

The creation of Page State and Actions follows a similar process as the current, with an intermediate dialog which sets the "type" of each:

![Screenshot 2023-08-01 at 1 26 55 AM](https://github.com/mui/mui-toolpad/assets/19550456/7b1ee35f-e68c-4ea8-9c4b-fce8bc9d1f52)


The actual editing/creation can happen in a proposed bottom panel (similar to the one outlined in https://github.com/mui/mui-toolpad/pull/2393#issue-1828023780) like so:

![Screenshot 2023-08-01 at 1 47 27 AM](https://github.com/mui/mui-toolpad/assets/19550456/a97481cc-ef21-499e-96c0-5fe2fc7a8e05)

### Other app-level pages

_Very low-level wireframes_

| Page | View |
| ----- | -------------------- |
| Theme| ![Screenshot 2023-08-01 at 2 08 03 AM](https://github.com/mui/mui-toolpad/assets/19550456/9785099f-2e1a-46fa-a0d8-a90b20da6b89) | 
| Custom components | ![Screenshot 2023-08-01 at 2 08 09 AM](https://github.com/mui/mui-toolpad/assets/19550456/2681fb16-b1ee-4dc4-a4f7-a0fd7f0c7b4c) | 
| Settings | ![Screenshot 2023-08-01 at 2 08 19 AM](https://github.com/mui/mui-toolpad/assets/19550456/e678288d-7e63-4d4b-b3db-1a01adca3f20) |

**Comment by prakhargupta1** [comment]
My thoughts on App level vs page level sidebar:

Toolpad supports building 1 app that can contain many pages. We don't have a Home view now so we need to decide where to put app-level menu items. I suggest we put them in the header bar as:

1. they will be used less and can be away from the main work area. 
2. code components can be inside components (like today) and would be better suited on the top bar as it is an editor feature that should be handy. Also removing drawer will give more horizontal space to the canvas/sidebar.
3. current theme is app-level and is definitely in the wrong place.
4. All of these can open in a pop-up/modal.

Having a page-level sidebar will declutter page building leading to a lower cognitive load of seeing everything together in the 4 vertical panels.

1. Page 
2. Page hierarchy explorer
3. Page state (All state-related info)
4. Actions
5. Page parameters

Note: Items like Connections, audit logs, org settings are out of the scope of this discussion and definitely can't be placed on the header bar.
